csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,457 @@
+
+      // check if we want to further propagate the replay on each operand
+      // separately
+      if (shouldPropagatePad(lhs, std::ref(frontier))) {
+        stack.emplace(bop, 0);
+      }
+      if (shouldPropagatePad(rhs, std::ref(frontier))) {
+        stack.emplace(bop, 1);
+      }
+    } else {
+      // Unrecognized operation stops propagation, push entry to frontier for
+      // replay
+      frontier.emplace(edge.val(), -1);
+    }
+  }
+
+  // NOTE: if we have not propagated pass the original pad_op input, that means
+  // no changes has been made at all.
+  if (frontier.count(pad_op->in()) != 0) {
+    return nullptr;
+  }
+
+  std::unordered_map<Val*, Val*> replacement_map;
+  // entries in frontier contains Val that we want to replay the PadOp on.
+  for (const auto& [pad_val, _] : frontier) {
+    auto pad_tv = pad_val->as<TensorView>();
+    const std::vector<IterDomain*> out_ids = TensorDomain::noReductions(
+        pad_op->out()->as<TensorView>()->getLogicalDomain());
+    // replay pad_op on frontier TVs assuming its output iter_type wouldn't
+    // change from the final output.
+    TensorView* new_out = replayConcretePad(
+        pad_tv, pad_op->value(), {pad_op->getPadWidths()}, out_ids);
+    replacement_map[pad_val] = new_out;
+  }
+
+  // reverse traversal the replay_sequence and update each input to use padded
+  // TVs.
+  std::reverse(replay_sequence.begin(), replay_sequence.end());
+  for (Expr* e : replay_sequence) {
+    if (e->isA<UnaryOp>()) {
+      Val* out = ops::newValLike(
+          replacement_map.at(e->input(0)), e->output(0)->getDataType().value());
+      Expr* padded_e = IrBuilder::create<UnaryOp>(
+          e->as<UnaryOp>()->getUnaryOpType(),
+          out,
+          replacement_map.at(e->input(0)));
+      replacement_map[e->output(0)] = padded_e->output(0);
+    } else if (e->isA<BinaryOp>()) {
+      std::vector<Val*> vals = {
+          replacement_map.at(e->input(0)), replacement_map.at(e->input(1))};
+      Val* out = ops::newOutputTV(vals, e->output(0)->getDataType().value());
+      Expr* padded_e = IrBuilder::create<BinaryOp>(
+          e->as<BinaryOp>()->getBinaryOpType(), out, vals[0], vals[1]);
+      replacement_map[e->output(0)] = padded_e->output(0);
+    } else {
+      NVF_ERROR(false, "expr type for propagation is not implemented");
+    }
+  }
+
+  // return the final replacement input to pad_op
+  return replacement_map.at(pad_op->in());
+}
+
+// This pass tries to push `PadOp`, which is part of `cat` (composed of `PadOp`s
+// + `CatOp`), further to its producers to avoid possible segmentation. As a
+// side effect, it replaces the `CatOp` with a series of pointwise add if any
+// mutation has been made for propagating `PadOp`
+void decomposeCatOp(Fusion* fusion) {
+  // TODO: verify that no dead branch is traversed in exprs.
+  std::vector<Expr*> exprs = fusion->exprs();
+
+  for (auto* cat : ir_utils::filterByType<CatOp>(exprs)) {
```

Comment:
To be constructive, I think there's an alternative to initially add all pads from all cats in the work list (whether a stack or a queue) and then push them upwards and combine if necessary. This can be more effective than the approach taken in this PR, and avoid the need for a separate #2490 to merge pads originated from different cats. Wdyt about that?

Reply:
I agree with you that it's more efficient a pass when we merge `pads` together before we propagate. Especially in the example you posted here.

```
z0 = cat([x, y0])
z1 = cat([x, y1])  // assume shape(y1)==shape(y0)
```

With our existing logic, we can't handle the horizontal cat in current implementation yet. But we will be able to do so if we try to merge `pad` before we push it out, assuming that we can prove they have identical `pad_widths` / `pad_value`.

I think if I swap the logic here to iterate from later nodes to earlier nodes and add a `merge pads` as part of the traversal logic, I can merge the second PR here as well and I can also ignore the tricky cases with a wild `padOp` could have negative padding.

Sounds like a good idea to me.

Reply:
Realized that this pattern is actually real in qkv_split.

i.e. cos/sin is used by cat on both q and k. i.e. if we need to propagate `pad` back, we need to handle this. Thanks for pointing this one out @wujingyue